lib/spack/spack/patch.py

```diff
@@ -199,11 +199,15 @@ def fetch(self, stage):
         fetcher = fs.URLFetchStrategy(self.url, fetch_digest,
                                       expand=bool(self.archive_sha256))
 
-        per_package_ref = os.path.join(
-            self.owner.split('.')[-1], os.path.basename(self.url))
+        # Patch url's can be ambiguous - The same patch name can have different
+        # versions that apply to different versions of the package. We append
+        # a bit of the hash to differentiate them. (Collision odds are exceedingly low
+        # given how few of these there should be, including b-day attack odds). 
+        name = '{0}-{1}'.format(os.path.basename(self.url), fetch_digest[:8])
```

Comment:
(very minor) where we use a subset of the hash we typically use the first 7 characters - can you change "8" to "7"?

Also I think the comment could be improved. Perhaps

> The same package can have multiple patches with the same name but with different contents, therefore apply a subset of the hash

also I don't think the qualification about collision is necessary (we do this in a few other spots without mentioning it), although I agree with your reasoning.

Reply:
Done.